File: src/main/java/com/ensemble/wardrobe/CategoryTaxonomy.java
package com.ensemble.wardrobe;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The wardrobe's single backend source of truth for the item category
 * taxonomy: the ordered list of taxonomy values, and the pure
 * {@link #normalize(String)} fallback that every save path runs a category
 * through so persisted data always lands on a taxonomy value or
 * {@code "Other"}.
 *
 * <p>{@code normalize} is critical logic (AGENTS.md / TESTING.md): it is
 * case- and whitespace-insensitive, maps a starter set of legacy/free-text
 * synonyms onto their taxonomy bucket, and <strong>never throws</strong> —
 * an unrecognized, blank, or {@code null} value simply buckets to
 * {@code "Other"}. This is the code-side guarantee behind the vision
 * tool-schema {@code enum}, which is only an advisory hint to the model
 * (see {@code AnthropicVisionModelClient.tagTool()}).
 *
 * <p>This constant is the backend half of a per-stack "two derived lists,
 * one intent" pair; the frontend defines the same taxonomy independently in
 * {@code frontend/src/lib/categoryTaxonomy.ts}. Cross-stack agreement between
 * the two is maintained <strong>by hand</strong> — no shared code, codegen, or
 * cross-stack test enforces it. A value or synonym added on one side only
 * (e.g. {@code romper → Dress}) silently diverges backend save-path
 * normalization from frontend edit/display-time normalization, and no test
 * fails. This is an explicitly accepted demo-scale trade-off: if you change
 * the taxonomy or synonym map here, mirror it in the frontend file in the
 * same change. Revisit with a shared JSON fixture or CI guard if the taxonomy
 * grows — see assumption A6.1 in
 * {@code docs/specs/18-spec-wardrobe-categories/18-assumptions-wardrobe-categories.md}.
 */
public final class CategoryTaxonomy {

	/** Ordered taxonomy values; {@code Other} is always last. */
	private static final List<String> VALUES =
		List.of("Jacket", "Top", "Bottom", "Dress", "Shoes", "Jewelry", "Accessory", "Other");

	private static final String OTHER = "Other";

	/**
	 * Case/whitespace-insensitive synonym → taxonomy-bucket map. Keys are
	 * lower-case; values are the canonical taxonomy strings. A starter map
	 * (assumption A1.12) covering the canonical values themselves plus
	 * representative legacy/free-text synonyms actually seen in wardrobe data.
	 */
	private static final Map<String, String> SYNONYMS = buildSynonyms();

	private CategoryTaxonomy() {
	}

	/** The ordered taxonomy values, {@code Other} last. Unmodifiable. */
	public static List<String> values() {
		return VALUES;
	}

	/**
	 * Maps a raw, possibly free-text category to a canonical taxonomy value.
	 * Case- and whitespace-insensitive; unrecognized, blank, or {@code null}
	 * input maps to {@code "Other"}. Never throws.
	 *
	 * @param rawCategory the raw stored/model-emitted/user-entered category
	 * @return a value from {@link #values()}, never {@code null}
	 */
	public static String normalize(String rawCategory) {
		if (rawCategory == null) {
			return OTHER;
		}
		String key = rawCategory.trim().toLowerCase(java.util.Locale.ROOT);
		if (key.isEmpty()) {
			return OTHER;
		}
		return SYNONYMS.getOrDefault(key, OTHER);
	}

	private static Map<String, String> buildSynonyms() {
		Map<String, String> map = new LinkedHashMap<>();
		put(map, "Jacket", "jacket", "coat", "blazer", "parka", "overcoat");
		put(map, "Top", "top", "shirt", "t-shirt", "tshirt", "tee", "sweater", "sweatshirt",
			"hoodie", "blouse", "polo", "knit");
		put(map, "Bottom", "bottom", "pants", "chinos", "jeans", "trousers", "shorts", "skirt", "leggings");
		put(map, "Dress", "dress", "gown");
		put(map, "Shoes", "shoes", "shoe", "sneakers", "boots", "loafers", "sandals", "heels", "footwear");
		put(map, "Jewelry", "jewelry", "jewellery", "necklace", "ring", "bracelet", "earrings", "watch");
		put(map, "Accessory", "accessory", "bag", "tote", "hat", "cap", "belt", "scarf",
			"sunglasses", "gloves", "tie");
		put(map, "Other", "other");
		return Map.copyOf(map);
	}

	private static void put(Map<String, String> map, String bucket, String... synonyms) {
		for (String synonym : synonyms) {
			map.put(synonym, bucket);
		}
	}
}

File: frontend/src/lib/categoryTaxonomy.ts
// Frontend single source of truth for the wardrobe item category taxonomy
// (assumption A1.4/A2.3): the ordered taxonomy values and the pure
// `normalizeCategory` fallback that both the `TagForm` `<select>` (Task 2.0) and
// the grid's grouping helper (Task 3.0) derive from. This mirrors the backend
// `com.ensemble.wardrobe.CategoryTaxonomy` constant one-for-one — the same
// ordered list and the same starter legacy-synonym map — but the two are
// intentionally separate, per-stack definitions (no shared codegen at this
// scale); cross-stack agreement is a review/test invariant.

/** Ordered taxonomy values; `Other` is always last. */
export const CATEGORIES = [
  'Jacket',
  'Top',
  'Bottom',
  'Dress',
  'Shoes',
  'Jewelry',
  'Accessory',
  'Other',
] as const

/** A canonical taxonomy value — one of {@link CATEGORIES}. */
export type Category = (typeof CATEGORIES)[number]

const OTHER: Category = 'Other'

/**
 * Case/whitespace-insensitive synonym → taxonomy-bucket map. Keys are
 * lower-case; values are canonical `Category` strings. A starter map
 * (assumption A1.12) covering the canonical values themselves plus
 * representative legacy/free-text synonyms actually seen in wardrobe data.
 */
const SYNONYMS: Record<string, Category> = buildSynonyms()

function buildSynonyms(): Record<string, Category> {
  const map: Record<string, Category> = {}
  const put = (bucket: Category, synonyms: string[]) => {
    for (const synonym of synonyms) {
      map[synonym] = bucket
    }
  }
  put('Jacket', ['jacket', 'coat', 'blazer', 'parka', 'overcoat'])
  put('Top', [
    'top',
    'shirt',
    't-shirt',
    'tshirt',
    'tee',
    'sweater',
    'sweatshirt',
    'hoodie',
    'blouse',
    'polo',
    'knit',
  ])
  put('Bottom', ['bottom', 'pants', 'chinos', 'jeans', 'trousers', 'shorts', 'skirt', 'leggings'])
  put('Dress', ['dress', 'gown'])
  put('Shoes', ['shoes', 'shoe', 'sneakers', 'boots', 'loafers', 'sandals', 'heels', 'footwear'])
  put('Jewelry', ['jewelry', 'jewellery', 'necklace', 'ring', 'bracelet', 'earrings', 'watch'])
  put('Accessory', [
    'accessory',
    'bag',
    'tote',
    'hat',
    'cap',
    'belt',
    'scarf',
    'sunglasses',
    'gloves',
    'tie',
  ])
  put('Other', ['other'])
  return map
}

/**
 * Maps a raw, possibly free-text category to a canonical taxonomy value.
 * Case- and whitespace-insensitive; unrecognized, blank, `null`, or
 * `undefined` input maps to `"Other"`. Never throws.
 */
export function normalizeCategory(raw: string | null | undefined): Category {
  if (raw == null) {
    return OTHER
  }
  const key = raw.trim().toLowerCase()
  if (key === '') {
    return OTHER
  }
  return SYNONYMS[key] ?? OTHER
}

/**
 * Singular→plural display-label map for wardrobe-grid section headers
 * (assumption A1.11). Explicit per value — not derived by string
 * concatenation — so irregular plurals (`Dress`→"Dresses") and unchanged
 * forms (`Shoes`→"Shoes", `Jewelry`→"Jewelry", `Other`→"Other") render
 * correctly.
 */
export const CATEGORY_LABELS: Record<Category, string> = {
  Jacket: 'Jackets',
  Top: 'Tops',
  Bottom: 'Bottoms',
  Dress: 'Dresses',
  Shoes: 'Shoes',
  Jewelry: 'Jewelry',
  Accessory: 'Accessories',
  Other: 'Other',
}

/**
 * The fixed display order for wardrobe-grid sections: taxonomy order with
 * `Other` always last (assumption A1.11). `CATEGORIES` already satisfies
 * this (`Other` is defined last), so this simply names that invariant for
 * callers that group by section.
 */
export function sectionOrder(): readonly Category[] {
  return CATEGORIES
}
